src/envoy/http/authn/http_filter.h

```diff
@@ -0,0 +1,55 @@
+/* Copyright 2018 Istio Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "src/envoy/http/authn/authn_store.h"
+
+#include "common/common/logger.h"
+#include "server/config/network/http_connection_manager.h"
+
+namespace Envoy {
+namespace Http {
+
+// The authentication filter.
+class AuthnFilter : public StreamDecoderFilter,
+                    public Logger::Loggable<Logger::Id::http> {
+ public:
+  AuthnFilter(Upstream::ClusterManager& cm, Auth::AuthnStore& store);
+  ~AuthnFilter();
+
+  // Http::StreamFilterBase
+  void onDestroy() override;
+
+  // Http::StreamDecoderFilter
+  FilterHeadersStatus decodeHeaders(HeaderMap& headers, bool) override;
+  FilterDataStatus decodeData(Buffer::Instance&, bool) override;
```

Comment:
Just curious, do you know if all these decodeXXX needed for the filter, or we just need to overwrite decodeHeaders?

Reply:
decodeXXX are pure virtual functions from the base class so they are implemented, which simply logs the events for diagnosis. We can probably place all authentication logic in decodeHeaders().

Reply:
Probably just put definition for decodeData & decodeTrailer here to keep it clear that they are not needed for authentication process.

Reply:
decodeData and decodeTrailer only does return FilterDataStatus::Continue.

Reply:
I know. I mean it simpler enough to just put it right here in the header, so we don't have to think about it. Anyway, it's just  my personal preference, no big deal.